### ISSUE TITLE
Feature request: write final image into custom file or device

### ISSUE THREAD

**Issue by jimis** [raspberrypi/rpi-imager#516]
I would like to download and prepare an image using `rpi-imager`, but **write it into a custom file**. 

The goal is to do further work on the image, before I copy it into one or several SD cards.

P.S. The help text shows an argument `<image file to write>` but this seems to serve a different purpose.
```
rpi-imager [--debug] [--version] [--repo <repository URL>] [--qm <custom qm translation file>] [--disable-telemetry] [<image file to write>]
-OR- rpi-imager --cli [--disable-verify] [--sha256 <expected hash>] [--debug] [--quiet] <image file to write> <destination drive device>
```

**Comment by lurch** [comment]
...although I'm not sure why you need to use `rpi-imager` to do that?
```
$ wget -q -O- https://releases.libreelec.tv/LibreELEC-RPi4.arm-10.0.3.img.gz | gunzip > test.img
$ ls -l test.img
-rw-rw-r-- 1 andrew andrew 575668224 Dec  4 16:52 test.img
```

**Comment by jimis** [comment]
Thanks! Unfortunately that doesn't seem to work in GUI mode.

 I want to use the GUI to do the basic setup (SSH, username, locale).

**Comment by maxnet** [comment]
>Thanks! Unfortunately that doesn't seem to work in GUI mode.

Sorry, prefer not to allow writing arbitrary devices there.

>I want to use the GUI to do the basic setup (network setup, enable SSH, set username, set locale).

If you write a customized SD card once, and steal the firstrun.sh from the FAT partition, Imager v1.7.4 do can apply the same firstrun.sh to later images.

```
rpi-imager --cli --first-run-script firstrun.sh [image file/url] [device to write to]
```

(option was added for automated testing recently.
Imager v1.7.4 is not released yet, but you do can compile it from source, or grab a [.deb here](https://www.maxnet.eu/tmp/174-fat/rpi-imager_1.7.4_amd64.deb)).

**Comment by jimis** [comment]
If the file `firstrun.sh` is all that is needed, would it be possible to only write that one somewhere?

**Comment by maxnet** [comment]
>If the file firstrun.sh is all that is needed, would it be possible to only write that one somewhere?

If it is a RPI OS based OS you can manually toss on the FAT partition:

- firstrun.sh
- a modified cmdline.txt that has ` systemd.run=/boot/firstrun.sh systemd.run_success_action=reboot systemd.unit=kernel-command-line.target ` on the same line as existing options.

If it's a Ubuntu based OS you need the following files on the FAT partitioninstead:

- user-data
- network-config

**Comment by jimis** [comment]
> If the file firstrun.sh is all that is needed, would it be possible to only write that one somewhere?

Is it possible for rpi-imager to write only this file somewhere, instead of writing the full image? This way I could customize this file and then manually add it to the installation image.

Thank you for the information, you have already provided very useful insight in what rpi-imager actually does.

**Comment by maxnet** [comment]
>Is it possible for rpi-imager to write only this file somewhere, instead of writing the full image?

No.
Although it does print it to console for debugging purposes if you start "rpi-imager" from terminal and "save" the custom settings.

(Do need to replace \n with a real new line in the output if using that output)

**Comment by JayGoldberg** [comment]
Respectfully this is really frustrating. I'm a ChromeOS user and I just wanted to configure `rpi-imager --cli --enable-writing-system-drives` which get's me a .bin image I can use with Chromebook Recovery Utility but of course this image is not customized because there aren't any CLI flags for setting Wi-FI or user, or SSH settings.

I cannot get a `firstrun.sh` from the GUI environment to try and make my own headless launch, because the GUI won't proceed until you have selected an output drive *which I don't want, anyway* and the GUI won't accept `--enable-writing-system-drives` or a .bin file destination at all.

Perhaps we can move the drive selection to the *end* of the configuration steps?

**Comment by lurch** [comment]
> Respectfully this is really frustrating. I'm a ChromeOS user and I just wanted to configure `rpi-imager --cli --enable-writing-system-drives` which get's me a .bin image I can use with Chromebook Recovery Utility

Respectfully that's an extremely niche use-case, and I don't expect that we'll be modifying Raspberry Pi Imager to support that.
But the good news is that Raspberry Pi OS Trixie (which we're still working on) will be using cloud-init rather than firstrun.sh, and although I don't know much about cloud-init myself I believe that it's more CLI-friendly?

**Comment by JayGoldberg** [comment]
> Chromebooks (running ChromeOS) dominate the K-12 education market, with some sources reporting 60% of global education market share. A significant percentage of students, possibly over half, use ChromeOS for their schooling.

If Raspberry Pi is an education tool (as purported in the original release and ethos about being the "next BBC Micro"), then ChromeOS compatibility is certainly important. Thus being able to flash/configure a Raspberry Pi image without custom binaries is a must.

`rpi-imager` **already** allows writing out to an image using the `--cli --enable-writing-system-drives` options but chooses to hide this from the user when using the GUI. But the GUI is the only way to configure options for a headless system!

The functionality *is already* in `rpi-imager`.

I was able to get around all of this--for Raspbery Pi OS--with [`custom.toml`](https://gist.github.com/lpenz/ef21bb38a7aa12ebde17fa719a8546b5), which is how this should all have worked in the first place. We should not need custom, proprietary, bloated binaries to flash a basic configured Raspberry Pi OS to a Raspberry Pi.

> cloud-init rather than firstrun.sh, and although I don't know much about cloud-init 

This is [not a legitimate planned, staffed goal at this time for Raspberry Pi devs](https://github.com/RPi-Distro/raspberrypi-sys-mods/issues/82). It has been discussed since October 2023. cloud-init is not happening soon.

For any educators out there using Pi in their classroom, look into a `custom.toml` which will likely work until cloud-init is figured out in 2027 or whatever

**Comment by lurch** [comment]
Can you please explain your use-case in more detail? I initially assumed that you were trying to flash an image to run on your Chromebook, but now I see that you're saying "flash a basic configured Raspberry Pi OS to a Raspberry Pi"?

> cloud-init is not happening soon.

Yes, it is 😉  https://github.com/raspberrypi/rpi-imager/pull/1201

**Comment by JayGoldberg** [comment]
### Headless install
The use case is pretty common: A person who doesn't have an extra keyboard or monitor to configure your Raspberry Pi (with Raspberry Pi OS). Or doesn't want to go through the trouble. They want a headless install.

You used to be able to do this on any platform, then add a `wpa_supplicant.conf` and `touch ssh` on the FAT32 boot partition. You could do this from basically any text editor on any OS.

### ChromeOS 
For classrooms, with students using ChromeOS, you'd need to have them all install Linux on ChromeOS (Crostini) to run `rpi-imager`. But that wouldn't work anyway because it won't see the SD card to write the output. They would need to write the image to a .bin that can be written by the Chromebook Recovery Utility (which is just `dd` for ChromeOS).

ChromeOS is indeed specific, but was captured by the `wpa_supplicant.conf` and `touch ssh` on the FAT32 boot partition flow. That was removed, pushing people to `rpi-imager` *which doesn't run on ChromeOS natively* and can't write to an SD card.

---
I understand the team wants to make a generally applicable flasher for multiple OSes. I'd posit that anyone wanting to use an OS other than Raspberry Pi OS is already "off the technical reservation" and is the edge case.

I'm glad that someone is working on cloud-init, I'm hoping that it allows a cross-platform headless configuration of a Raspberry Pi without extra software.

**Comment by lurch** [comment]
I've never used a Chromebook or ChromeOS myself. But just to clarify, the use-case is that you have a classroom full of Chromebooks and Raspberry Pis, and you install a Linux environment within ChromeOS to run RPi Imager, and then use that to download and configure a Raspberry Pi OS image, writing the result to an `rpios.bin` file (within the Linux environment), and then back in "native" ChromeOS you use Chromebook Recovery Utility to write that `rpios.bin` file to a microSD card, which you then use to boot up the RPi, and then connect from the Chromebook to the RPi using SSH?
Yeah, that's not really a workflow that we have a good solution for at the moment. See also https://github.com/raspberrypi/rpi-imager/issues?q=is%3Aissue%20chromeos  Perhaps @tdewey-rpi might have some good ideas when he's back from vacation.

There used to be http://pinet.org.uk/ specifically for classroom environments, but unfortunately that's no longer maintained.

**Comment by JayGoldberg** [comment]
Yes, that's roughly what it is about. But as per OP, the key factor is "being to be able to do something with the customized .bin file" whether that's flash it with `dd` later or modify the image. As of this time, `rpi-imager` refuses to write to a .bin file in GUI mode.

Ideally the `rpi-imager` will not be required at all--which I know is against this project's interests--you can `dd` an image and customize it through text files on the boot partition. I'm hoping this is what cloud-init allows but I have to do more research.

**Comment by tdewey-rpi** [comment]
I'm not immediately averse to the concept - but in the past I've suggested that users looking for a more flexible customisation scheme cut out the middle man and go directly to Rpi-image-gen.

Is there a reason that wouldn't be viable for you, @JayGoldberg ?

**Comment by JayGoldberg** [comment]
I looked at `rpi-image-gen` but cannot find features to set Wi-Fi config (though I see SSH and user1 config options). It seems lower level than that. I'll keep hoping `rpi-imager` will someday allow users to redirect output to a disk file, as per UNIX conventions.

To me the ideal solution is to have the `custom.toml` method continue to work, but for other OSes that `rpi-imager` supports, a filepath for output would be great. Thanks.

**Comment by tdewey-rpi** [comment]
`cloud-init` has shipped in Raspberry Pi OS, and Imager allows selection of system drives (including virtual disks).

Marking this one complete.
